src/shopify/graphql/fetchProduct.ts

```diff
@@ -1,26 +1,126 @@
+        query: getProductsByHandles,
+        variables: {
+          language: window.Shopify?.locale?.toUpperCase() || "EN",
+          country: window.Shopify?.country || "US",
+          query: queryString,
+          first: handles.length
+        }
+      })
+    })
+
+    if (!response.ok) {
+      const error = new Error(`Failed to fetch product data: ${response.status} ${response.statusText}`)
+      for (const requests of requestsMap.values()) {
+        for (const request of requests) {
+          request.reject(error)
+        }
+      }
+      return
+    }
+
+    const responseData = await response.json()
+
+    // Create a map of handle to product
+    const productsByHandle = new Map<string, ShopifyProduct>()
+    if (responseData.data?.products?.nodes) {
+      for (const productNode of responseData.data.products.nodes) {
+        const product = flattenResponse({ data: { product: productNode } })
+        productsByHandle.set(productNode.handle, product)
+      }
+    }
+
+    // Resolve or reject each request
+    for (const handle of handles) {
+      const requests = requestsMap.get(handle)
+      if (!requests) continue
+
+      const product = productsByHandle.get(handle)
+      if (product) {
+        for (const request of requests) {
+          request.resolve(product)
+        }
+      } else {
+        const error = new Error(`Product not found: ${handle}`)
+        for (const request of requests) {
+          request.reject(error)
+        }
+      }
+    }
+  }
+
+  async function flush() {
+    state.scheduledFlush = null
+
+    // Get unique handles from pending requests
+    const handles = Array.from(state.pendingRequests.keys())
+    const requestsMap = new Map(state.pendingRequests)
+    state.pendingRequests.clear()
+
+    if (handles.length === 0) {
+      return
+    }
+
+    try {
+      await fetchBatch(handles, requestsMap)
+    } catch (error) {
+      // Reject all pending requests with the error
+      for (const requests of requestsMap.values()) {
+        for (const request of requests) {
+          request.reject(error instanceof Error ? error : new Error(String(error)))
+        }
+      }
+    }
+  }
+
+  function request(handle: string): Promise<ShopifyProduct> {
+    return new Promise((resolve, reject) => {
+      // Get or create array for this handle
+      const requests = state.pendingRequests.get(handle) || []
+      requests.push({ handle, resolve, reject })
+      state.pendingRequests.set(handle, requests)
+
+      // Schedule flush if not already scheduled
+      if (state.scheduledFlush === null) {
+        state.scheduledFlush = requestAnimationFrame(() => flush())
       }
     })
-  })
-  if (!response.ok) {
-    throw new Error(`Failed to fetch product data: ${response.status} ${response.statusText}`)
   }
-  const responseData = await response.json()
-  return flattenResponse(responseData)
-})
+
+  return { request }
+}
+
+const batcher = createProductBatcher()
```

Comment:
as there is only a single batcher the createProductBatcher content can be flattened to this module